doc/src/manual/methods.md

```diff
@@ -515,6 +515,225 @@ julia> wait(schedule(t, 1))
+object with some change made to the layout of the type.
+For instance, you might have some sort of abstract array with an arbitrary element type
+and want to write the computation on it with a specific element type.
+
+Another implication of the common mistake discussed above mentioned above,
+is that we must implement a method for each type that describes how to compute this type transform.
+There is no general transform of one subtype into another subtype with a different parameter.
+(Quick review: do you see why this is?)
+
+The subtypes of `AbstractArray` typically implement two methods for expressing this:
+A method to convert the input to a subtype of a specific `AbstractArray{T, N}` abstract type;
+and a method to make a new uninitialized matrix with a specific element type.
+Sample implementations of these can be found in the standard library.
+Here is a basic example usage of them:
+
+```julia
+input = convert(AbstractArray{Eltype}, input)
+output = similar(Eltype, input)
+```
+
+As an extension of this, in cases where the algorithm wants to mutate the input array,
+[`convert`](@ref) is insufficient as the return value may alias the original input.
+Combining [`similar`](@ref) (to make the output array) and [`copy!`](@ref) (to fill it with the input data)
+is a generic way to express the requirement for a mutable copy of the input argument:
+
+```julia
+copy_with_eltype(Eltype, input) = copy!(similar(Eltype, input), input)
+```
+
+### Iterated dispatch
+
+In order to dispatch a multi-level parametric argument list,
+often it is best to separate each level of dispatch into separate functions.
+This may sound similar in approach to single-dispatch, but as we shall see below, it is still more flexible.
+
+For example, trying to dispatch on the element-type of an array will often run into ambiguous situations.
+Instead, commonly code will dispatch first on the container type,
+then recurse down to a more specific method based on eltype.
+In most cases, the algorithms lend themselves conveniently to this hierarchical approach,
+while in other cases, this rigor must be resolved manually.
+This dispatching branching can be observed, for example, in the logic to sum two matrices:
+
+```julia
+# First dispatch selects the map algorithm for element-wise summation.
++(a::Matrix, b::Matrix) = map(+, a, b)
+# Then dispatch handles each element and selects the appropriate
+# common element type for the computation.
++(a, b) = +(promote(a, b)...)
+# Once the elements have the same type, they can be added.
+# For example, via primitive operations exposed by the processor.
++(a::Float, b::Float) = Core.add(a, b)
+```
+
+### Trait-based dispatch
+
+A natural extension to the iterated dispatch above is to add a layer to
+method selection that identifies classes of types.
+We could do this by writing out a `Union` of the types with similar characteristics.
+But then this list would not be extensible.
+However, by expressing the property as a "trait",
```

Comment:
_by defining "traits"_? "trait" is synonymous to a "property", and the latter has not been introduced before